File: playeranimations.go
package main

import "github.com/veandco/go-sdl2/sdl"

type spriteCoord struct {
	Y, X uint8
}

var (
	idleCoords = []spriteCoord{
		{0, 0}, {0, 0}, {0, 1}, {0, 1}, {0, 2}, {0, 2}, {0, 3}, {0, 3},
	}
	runningCoords = []spriteCoord{
		{1, 1}, {1, 2}, {1, 3}, {1, 4}, {1, 5}, {1, 6},
	}
	jumpingCoords = []spriteCoord{
		{2, 0}, {2, 1}, {2, 2}, {2, 3}, {2, 4},
	}
	fallingCoords = []spriteCoord{
		{3, 1}, {3, 2},
	}
	wallSlidingCoords = []spriteCoord{
		{11, 2}, {11, 3},
	}
	wallJumpingCoords = []spriteCoord{
		{11, 0}, {11, 1},
	}
)

var dispatch map[PlayerState][]spriteCoord = map[PlayerState][]spriteCoord{
	Idle:        idleCoords,
	Running:     runningCoords,
	Jumping:     jumpingCoords,
	Falling:     fallingCoords,
	WallSliding: wallSlidingCoords,
	WallJumping: wallJumpingCoords,
}

func (player *Player_t) setTextureArea() {
	spriteCoordArray := dispatch[player.State]
	spriteCoord := spriteCoordArray[player.Frame/StepsPerFrame]

	player.TextureArea = &sdl.Rect{
		X: int32(spriteCoord.X) * CharacterWidth,
		Y: int32(spriteCoord.Y) * CharacterHeight,
		W: CharacterWidth,
		H: CharacterHeight,
	}

	player.Frame++
	player.Frame %= uint8(len(spriteCoordArray) * StepsPerFrame)
}

File: cs/PlayerAnimations.cs
using System;
using SDL2;

namespace shootingame
{
    class PlayerAnimations
    {
        private static readonly SDL.SDL_Point[] idleCoords = {
            SDLFactory.MakePoint(0, 0), SDLFactory.MakePoint(0, 0),
            SDLFactory.MakePoint(0, 1), SDLFactory.MakePoint(0, 1),
            SDLFactory.MakePoint(0, 2), SDLFactory.MakePoint(0, 2),
            SDLFactory.MakePoint(0, 3), SDLFactory.MakePoint(0, 3)
        };
        private static readonly SDL.SDL_Point[] runningCoords = {
            SDLFactory.MakePoint(1, 1), SDLFactory.MakePoint(1, 2),
            SDLFactory.MakePoint(1, 3), SDLFactory.MakePoint(1, 4),
            SDLFactory.MakePoint(1, 5), SDLFactory.MakePoint(1, 6)
        };
        private static readonly SDL.SDL_Point[] jumpingCoords = {
            SDLFactory.MakePoint(2, 0), SDLFactory.MakePoint(2, 1),
            SDLFactory.MakePoint(2, 2), SDLFactory.MakePoint(2, 2),
            SDLFactory.MakePoint(2, 3), SDLFactory.MakePoint(2, 4)
        };
        private static readonly SDL.SDL_Point[] fallingCoords = {
            SDLFactory.MakePoint(3, 1), SDLFactory.MakePoint(3, 2)
        };
        private static readonly SDL.SDL_Point[] wallSlidingCoords = {
            SDLFactory.MakePoint(11, 2), SDLFactory.MakePoint(11, 3)
        };
        private static readonly SDL.SDL_Point[] wallJumpingCoords = {
            SDLFactory.MakePoint(11, 0), SDLFactory.MakePoint(11, 1)
        };

        private static readonly SDL.SDL_Point[][] spriteCoord = {
            idleCoords, runningCoords, jumpingCoords, fallingCoords, wallSlidingCoords, wallJumpingCoords
        };

        public static void SetTextureArea(Player player)
        {
            var spriteCoordArray = spriteCoord[(int)player.State];
            SDL.SDL_Point coord = spriteCoordArray[player.Frame / Const.StepsPerFrame];

            player.TextureArea = SDLFactory.MakeRect(
                x: coord.x * Const.PlayerSpriteWidth,
                y: coord.y * Const.PlayerSpriteHeight,
                w: Const.PlayerSpriteWidth,
                h: Const.PlayerSpriteHeight
            );

            ++player.Frame;
            player.Frame %= (uint)(spriteCoordArray.Length * Const.StepsPerFrame);
        }
    }
}